include/onnxruntime/core/session/onnxruntime_c_api.h

```diff
@@ -1382,6 +1382,37 @@ struct OrtApi {
                   _In_ const void* model_data, size_t model_data_length,
                   _In_ const OrtSessionOptions* options, _Inout_ OrtPrepackedWeightsContainer* prepacked_weights_container,
                   _Outptr_ OrtSession** out);
+
+  /**
+   * Use this API to create the configuration of a TensorRT Execution Provider.
+   * \param out - pointer to the pointer of TensorRT EP provider options instance.
+   */
+  ORT_API2_STATUS(CreateTensorRTProviderOptions, _Outptr_ OrtTensorRTProviderOptions** out);
+
+  /**
+  * Use this API to set appropriate configuration knobs of a TensorRT Execution Provider.
+  * \tensorrt_provider_options - OrtTensorRTProviderOptions instance
+  * \provider_options_keys - array of UTF-8 null-terminated string for provider options keys
+  * \provider_options_values - array of UTF-8 null-terminated string for provider options values
+  * \num_keys - number of keys
+  */
+  ORT_API2_STATUS(UpdateTensorRTProviderOptions, _Inout_ OrtTensorRTProviderOptions* tensorrt_provider_options,
+                 _In_reads_(num_keys) const char* const* provider_options_keys,
+                 _In_reads_(num_keys) const char* const* provider_options_values,
+                 _In_ size_t num_keys);
+
+  /**
+  * Get configuration of a TensorRT Execution Provider.
+  * \param allocator - a ptr to an instance of OrtAllocator obtained with CreateAllocator() or GetAllocatorWithDefaultOptions()
+  *                      the specified allocator will be used to allocate continuous buffers for output strings and lengths.
+  * \param ptr - is a UTF-8 null terminated string allocated using 'allocator'. The caller is responsible for using the same allocator to free it.
+  */
+  ORT_API2_STATUS(GetTensorRTProviderOptions, _Inout_ OrtAllocator* allocator, _Outptr_ char** ptr);
```

Comment:
* It's not clear what this API does. Is it returning the serialized options str for the instance of tensor RT provider appended to the session? or is it returning the default options?
* Also, how does the user know the serialization format of the options?
* What is an example usage of this API?

Reply:
Re:
1. No, the returning serialized options string of this API is not for the TensorRT instance to be appended to the session. The returning serialized string is for user to see the default options as well as to check whether the options have been correctly configured after they call UpdateTensorRTProviderOptions.

2. The serialized options string is only for reference not for configuration. So I think they only need to know the returning is a serialized string which has been documented before the function declaration. 

3. When using this C API, SessionOptionsAppendExecutionProvider_TensorRT(_In_ OrtSessionOptions* options, _In_ const OrtTensorRTProviderOptions* tensorrt_options), it's possible that user only needs to change one option of tensorrt_options and keep other options as default values. In this case, they can call this API to get the default values instead of searching for code to get default value.
Also, this API can be used in unit test.

Reply:
Please check the new commit which I made for addressing default options.

Reply:
Since you want to return a serialized representation of OrtTensorRTProviderOptions, why don't you take OrtTensorRTProviderOptions as input to this API? The CreateTensorRTProviderOptions fn will return the default instance which can then be fed to GetTensorRTProviderOptions to get the string repr. And later if the user updates the default instance, he/she can pass the same instance to GetTensorRTProviderOptions to get the updated string repr to verify.
Also, I would rename GetTensorRTProviderOptions to GetTensorRTProviderOptionsAsString() or TensorRTProviderOptionsToString().

Reply:
It sounds reasonable to me. I will rename this API and make it purely functional.

At first, I made this API include the state of OrtTensorRTProviderOptions in TRT EP is that I want the state to keep status of what TRT provider options is being configured with TensorrtProviderFactory.
But later I found that in session options it also keeps provider options info as below.
https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/framework/execution_provider.h#L121
https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cuda/cuda_execution_provider.h#L77
So to make it simple and clear, I'll remove the internal state in TRT EP, aka the variable trt_options_.